inbox/mailsync/service.py

```diff
@@ -368,7 +381,11 @@ def stop_sync(self, account_id):
         with self.semaphore:
             self.log.info("Stopping monitors", account_id=account_id)
             if account_id in self.email_sync_monitors:
-                self.email_sync_monitors[account_id].kill()
+                email_sync_monitor = self.email_sync_monitors[account_id]
+                if email_sync_monitor.delete_handler:
+                    email_sync_monitor.delete_handler.kill()
+                email_sync_monitor.sync_greenlet.kill(block=False)
+                email_sync_monitor.join()
```

Comment:
we were previously killing the main email_sync_monitor greenlet, without killing their children which would become runaway greenlets. Because of that the cleanup logic was not called properly because the whole process would exit before those runaway greenlets ended.

Reply:
@mrhiggi-close 

The whole thing is kind of involved and probably needs a refactor or an explanatory comment at least. My bad excuse is that I did not write this concurrency mess myself :-) I will follow up with a PR that explains this better.

`email_sync_monitor.sync_greenlet` is the [child](https://github.com/closeio/sync-engine/blob/4c7e0d7f6cdace00eac837544c454e2fce165604/inbox/mailsync/backends/base.py#L69-L75) `email_sync_monitor` greenlet. `email_sync_monitor` greenlet [joins](https://github.com/closeio/sync-engine/blob/4c7e0d7f6cdace00eac837544c454e2fce165604/inbox/mailsync/backends/base.py#L77) `email_sync_monitor.sync_greenlet`, and that's the only thing it waits for.

So we can `email_sync_monitor.sync_greenlet.kill(block=False)` without blocking because `email_sync_monitor` starts exiting at this point as it was only joining that other greenlet. Then we can just `email_sync_monitor.join()` becuause it won't block on anything anymore. It's also important not to `email_sync_monitor.kill()` because this will possibly prevent it from running the [cleanup](https://github.com/closeio/sync-engine/blob/4c7e0d7f6cdace00eac837544c454e2fce165604/inbox/mailsync/backends/base.py#L79-L80) step here. The cleanup does database queries and they can be interrupted in the middle since gevent can raise `GreenletKill` exception at any point we are doing I/O.

It's terrible. I know.